src/script/standard.cpp

```diff
@@ -54,6 +55,26 @@ bool Solver(const CScript& scriptPubKey, txnouttype& typeRet, vector<vector<unsi
 
         // Sender provides N pubkeys, receivers provides M signatures
         mTemplates.insert(make_pair(TX_MULTISIG, CScript() << OP_SMALLINTEGER << OP_PUBKEYS << OP_SMALLINTEGER << OP_CHECKMULTISIG));
+
+        // HTLC where sender requests preimage of a hash
+        {
+            const opcodetype accepted_hashers[] = {OP_SHA256, OP_RIPEMD160};
+            const opcodetype accepted_timeout_ops[] = {OP_CHECKLOCKTIMEVERIFY, OP_CHECKSEQUENCEVERIFY};
+
+            BOOST_FOREACH(opcodetype hasher, accepted_hashers) {
+                BOOST_FOREACH(opcodetype timeout_op, accepted_timeout_ops) {
+                    mTemplates.insert(make_pair(TX_HTLC, CScript()
+                        << hasher << OP_ANYDATA << OP_EQUAL
```

Comment:
In the "ELSE" case any relay node may replace the top stack with anything without invalidating the tx. See related discussion at https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013036.html